contracts/src/CrossChainApplications/NativeTokenBridge/NativeTokenDestination.sol

```diff
@@ -0,0 +1,167 @@
+        bytes32 sourceChainID_
+    ) {
+        currentChainID = WarpMessenger(WARP_PRECOMPILE_ADDRESS)
+            .getBlockchainID();
+
+        if (teleporterMessengerAddress == address(0)) {
+            revert InvalidTeleporterMessengerAddress();
+        }
+        teleporterMessenger = ITeleporterMessenger(teleporterMessengerAddress);
+
+        if (sourceContractAddress_ == address(0)) {
+            revert InvalidSourceContractAddress();
+        }
+        sourceContractAddress = sourceContractAddress_;
+
+        if (sourceChainID_ == currentChainID) {
+            revert CannotBridgeTokenWithinSameChain();
+        }
+        sourceChainID = sourceChainID_;
+    }
+
+    /**
+     * @dev See {ITeleporterReceiver-receiveTeleporterMessage}.
+     *
+     * Receives a Teleporter message and routes to the appropriate internal function call.
+     */
+    function receiveTeleporterMessage(
+        bytes32 nativeChainID,
+        address nativeBridgeAddress,
+        bytes calldata message
+    ) external nonReentrant {
+        // Only allow the Teleporter messenger to deliver messages.
+        if (msg.sender != address(teleporterMessenger)) {
+            revert Unauthorized();
+        }
+        // Only allow messages from the partner chain.
+        if (nativeChainID != sourceChainID) {
+            revert InvalidSourceChain();
+        }
+        // Only allow the partner contract to send messages.
+        if (nativeBridgeAddress != address(this)) {
+            revert InvalidPartnerContractAddress();
+        }
+
+        (address recipient, uint256 amount) = abi.decode(
+            message,
+            (address, uint256)
+        );
+
+        // Calls NativeMinter precompile through INativeMinter interface.
+        _nativeMinter.mintNativeCoin(recipient, amount);
+        emit MintNativeTokens(recipient, amount);
+    }
+
+    /**
+     * @dev See {INativeTokenMinter-bridgeTokens}.
+     *
+     * Requirements:
+     *
+     * - `msg.value` must be greater than the fee amount.
+     */
+    function transferToSource(
+        address recipient,
+        address feeTokenContractAddress,
+        uint256 feeAmount
+    ) external payable nonReentrant {
+        // The recipient cannot be the zero address.
+        if (recipient == address(0)) {
+            revert InvalidRecipientAddress();
+        }
+
+        // Lock tokens in this bridge instance. Supports "fee/burn on transfer" ERC20 token
+        // implementations by only bridging the actual balance increase reflected by the call
+        // to transferFrom.
+        if (feeAmount > 0) {
+            uint256 adjustedAmount = SafeERC20TransferFrom.safeTransferFrom(
+                IERC20(feeTokenContractAddress),
+                feeAmount
+            );
+
+            // Ensure that the adjusted amount is greater than the fee to be paid.
+            if (adjustedAmount <= feeAmount) {
+                revert InsufficientAdjustedAmount(adjustedAmount, feeAmount);
```

Comment:
I don't think this check should be here. The `adjustedAmount` in this case could be less than the `feeAmount` specified, if it is a fee/burn on transfer token.